bindings/python/iree/runtime/array_interop.py

```diff
@@ -143,6 +143,20 @@ def astype(self, dtype, casting="unsafe", copy=True):
     host_ary = self.to_host()
     return host_ary.astype(dtype, casting=casting, copy=copy)
 
+  def reshape(self, *args):
+    # TODO(scotttodd): add a native impl with a new buffer_view of the same data
+    # TODO(scotttodd): return DeviceArray instead of host ndarray?
+    host_ary = self.to_host()
```

Comment:
reshape making a host copy for a metadata-only change? seems bad - major footgun to add a reshape and have your model run several times slower if you then passed the reshaped thing into an invocation and transfer everything over the PCI bus

Reply:
I agree, but I'm having a hard time tracing through the Python / pybind / C code to see if there is a way to make such a change really be "metadata-only" at this level, without more refactoring.

https://github.com/google/iree/blob/31d1d04ff4108a6db63f2a242996383a5f4269e8/bindings/python/iree/runtime/array_interop.py#L198-L206

https://github.com/google/iree/blob/31d1d04ff4108a6db63f2a242996383a5f4269e8/bindings/python/iree/runtime/hal.cc#L80-L82

Reply:
At least implementing `reshape()` like this will let users rely on it today. Without at least an inefficient stub/alias, they'll be forced to use explicit `to_host()` calls that will be out of our reach to optimize later.

Reply:
Yeah, this is why I didn't implement some of these: there are much better ways to do it but they are a bit out of reach for now. I do agree with the rationale, though, that if we make users put the barrier in their code, we won't be able to optimize it later.

Reply:
Would you mind adding a test for `np.reshape(ary)`? I think that would fail with this set up.

I think the canonical way to do this is to add an implementation to _DEVICE_HANDLED_FUNCTIONS (see the definition for np.shape below). Then you add a stub here:

```
def reshape(self, *args):
  return np.reshape(*args)
```

You'll see various levels of meta-programming in the wild to satisfy both. Try just adding this as a _DEVICE_HANDLED_FUNCTION without the reshape instance method and see if that works as is (I think there is some mixin magic).

Reply:
reshape should do a copy wherever the buffer is (host->host, device->device) - the thing that sketches me out here is the semantics change - e.g. have a device buffer I can't map, reshape it, then suddenly can map it. If we preserve the memory properties (type/usage/etc) then we can eventually pipeline things (reshape enqueues a copy command that isn't waited on until the contents are used or it's passed in as a wait to the invocation, etc).

[ndarray.resize](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.resize.html) is really what people should probably be using, though, as when the element counts match for us that's just recreating a buffer view wrapping the same buffer and effectively free besides a 64 byte realloc of the metadata. If element counts differ it goes down an alloc+copy(+fill) path, but is still something we could pipeline.

Reply:
To Stella's comment: Implemented `np.reshape` using the `_device_implements` decorator.

and to Ben's: I'm also looking into a faster path by exposing `iree_hal_buffer_view_buffer()` -> create a new buffer_view from the existing buffer_view's buffer, but with a new shape.

Reply:
I'm still fumbling through the pybind code to write a more efficient `reshape()` implementation: https://github.com/ScottTodd/iree/commit/c135bee7e8f9deee091bebeef8684ffc91df8eb1. Can we proceed with this PR without that, or should I remove the slow version of reshape() and add explicit .to_host calls to the affected Colab notebooks?